### ISSUE TITLE
Cache-mode and upstream Cache-Control

### ISSUE THREAD

**Issue by mnot** [whatwg/fetch#722]
[HTTP-network-or-cache-fetch](https://fetch.spec.whatwg.org/#http-network-or-cache-fetch) says:

> 12. If httpRequest’s cache mode is "default" and httpRequest’s header list contains `If-Modified-Since`, `If-None-Match`, `If-Unmodified-Since`, `If-Match`, or `If-Range`, then set httpRequest’s cache mode to "no-store".

The following text was added to fix #201, #231 and #232 (AFAICT):

> 13. If httpRequest’s cache mode is "no-cache" and httpRequest’s header list does not contain `Cache-Control`, then append `Cache-Control`/`max-age=0` to httpRequest’s header list.
>
> 14. If httpRequest’s cache mode is "no-store" or "reload", then:
> 
> >  1. If httpRequest’s header list does not contain `Pragma`, then append `Pragma`/`no-cache` to httpRequest’s header list.
> >
> >  2. If httpRequest’s header list does not contain `Cache-Control`, then append `Cache-Control`/`no-cache` to httpRequest’s header list.

This means that in several situations when the developer is trying to influence the local cache behaviour, it triggers request `Pragma` and `Cache-Control` headers that effectively disable upstream caches, even though this might not be desired.

In particular, setting `If-Modified-Since`, `If-None-Match` and other validating headers doesn't need to disable upstream caches; it's an entirely reasonable expectation for the upstream cache to be able to serve the request as normal. As specified, this is only an incentive for upstream caches to ignore these request directives, which isn't good for interop.

Even without validation, a developer might wish to bypass the browser cache with `no-store` but still use upstream caches (perhaps because they have a reliable out-of-band way to invalidate those caches, e.g., in a CDN).

I think the best solution here would be to remove (13) and (14), or to provide a separate control over these behaviours (e.g., an init flag to turn it off), so that developers have as much flexibility as they need; as specified, fetch() is making a decision for them. 

If not, at a minimum the [documentation for cache mode](https://fetch.spec.whatwg.org/#concept-request-cache-mode) needs to be updated to indicate that it is intended to influence not only the local cache, but also upstream caches.

P.S. I've tested to see if these headers are actually added, by making `no-store` requests. From what I can see:

* Chrome doesn't append `Pragma: no-cache` or `Cache-Control: no-cache`
* Firefox only appends `Pragma: no-cache`, and ignores the "If httpRequest's header list does not contain `Pragma`" condition (i.e, it appends it even if `Pragma` is present)
* Safari appends both if they aren't already present

**Comment by mnot** [comment]
Update: it looks like Safari does *not* follow the condition for setting `Pragma` and `Cache-Control` if the request headers contain `Range`. Weird.

**Comment by annevk** [comment]
cc @ehsan

**Comment by mnot** [comment]
@annevk should I do a PR?

**Comment by annevk** [comment]
This was added in 010dd7ad85d9bb893c7bbb03e2cbb800068da18e.

A PR would be good. It'd be nice if implementers chimed in, but maybe once we have a PR and tests that'll be easier. It does seem like this is poorly supported and if it's not exactly what we want anyway we might as well try to rid ourselves of it.

I'll try to ensure any cache changes are run by HTTP cache experts going forward as well.

**Comment by wanderview** [comment]
FWIW, I think this will be tricky for firefox to implement currently.  This stuff gets set deep in the networking stack:

https://searchfox.org/mozilla-central/rev/5dbfd833bbb114afe758db4d4bdbc5b13bcc33ef/netwerk/protocol/http/nsHttpChannel.cpp#1187-1200

AFAICT there is no higher level flag to bypass just the local cache.  This can probably be changed (if mozilla networking folks agree), but I think this is partly how we ended up here.

However, consider:

1. The browser wants to bypass upstream caches for some loads.  Like ctrl-r.
2. A service worker can intercept these loads and may proxy them via `evt.respondWith(fetch(evt.request))`.
3. This change makes `evt.request.cache` no longer reflect the intention of (1).

How do we maintain our goal of "service worker proxies requests don't lose functionality"?

**Comment by mayhemer** [comment]
So, you actually want to revert #201, #231 and #232 and give the control over transparent proxies to developers?

Note that Gecko has a LOAD_BYPASS_LOCAL_CACHE flag [1].  Setting that flag will bypass opening the local cache entry and won't set any Cache-control request headers.  But, will also refresh DNS (not sure why this has been designed that way..)


[1] https://searchfox.org/mozilla-central/rev/847b64cc28b74b44c379f9bff4f415b97da1c6d7/netwerk/base/nsICachingChannel.idl#96-101

**Comment by mnot** [comment]
s/transparent proxies/all upstream caches/ -- yes.

**Comment by ricea** [comment]
Chrome's network stack has similar issues to those described by @wanderview in https://github.com/whatwg/fetch/issues/722#issuecomment-413559634. The code that adds the headers based on the flags is deep inside the network stack: https://cs.chromium.org/chromium/src/net/http/http_network_transaction.cc?type=cs&q=HttpNetworkTransaction::BuildRequestHeaders&g=0&l=1090.

I think it's easier for Chrome to add more headers to converge with other browsers than it is to remove headers, so from Chrome's point of view it would be better to retain steps 13 and 14.

**Comment by mnot** [comment]
Implementer convenience is not what we should be optimising for.

AIUI Fetch is a low-level HTTP API -- as per EWM -- and that means it should not impose its views on its users, nor preclude any (safe, sane) functionality. 

Talking to @mcmanus, I was lead to believe this wasn't necessarily a big deal to correct; Patrick, any comment?

**Comment by mnot** [comment]
#852 seems to have added a `prevent no-cache cache-control header modification flag` that at least offers control over `no-cache`, even if the default hasn't changed and it doesn't address `no-store`.

**Comment by mnot** [comment]
Fetch also says:

> If header list contains `If-Modified-Since`, `If-None-Match`, `If-Unmodified-Since`, `If-Match`, or `If-Range`, fetch will set cache mode to "no-store" if it is "default".

This means that if the developer's intent is to do validation manually, Fetch will effectively disable that by appending `Cache-Control: no-cache`.

**Comment by mnot** [comment]
I created some tests for the current spec; it looks like Safari passes all, Chrome has one error (it sends `Cache-Control: no-cache` instead of `max-age=0` when the mode is `no-cache`) and Firefox has a number of errors.

I can get the results I need if the spec is followed (not changing `Cache-Control` when it's already set). I still think the right thing for Fetch to be doing is _not_ setting headers based upon what it thinks the developer wants.

**Comment by annevk** [comment]
Thanks for writing those @mnot. I guess this is largely a question for @whatwg/http. And perhaps it's something that could be discussed in the HTTP WG given where most of this logic sits in implementations?

**Comment by mnot** [comment]
This is really an API / client decision; if you want to be 'helpful' to the user, you try to guess what they intend (as Fetch currently does), while if you want to expose the complete capabilities of the protocol, you don't.

**Comment by pmeenan** [comment]
FWIW, this issue persists today and popped up with a situation where we are fetching a LOT of data (several gigabytes) and storing it in application-controlled data stores (indexdb in this case).  We want to bypass the local HTTP cache to avoid double-storing gigabytes worth of data but we don't want to punch a hole through intermediary caches.

cache: "no-store" on requestInit does exactly what we want in Chrome and Firefox today (doesn't append Cache-Control headers) but Safari adds the Cache-Control headers.

Presumably, the app could add the network headers directly if they wanted to poke holes in the network-level caches and `cache` could control the local caching behaviors.

**Comment by annevk** [comment]
I think it's reasonable to change the specification here. Also for "`reload`" presumably?

@valenting @youennf thoughts? @jgraham said that this was part of Interop 2023 in https://github.com/web-platform-tests/wpt/pull/37986 so deciding on what to do here would be good.

**Comment by mnot** [comment]
Yes, I think the proposal is to remove what are now (17) and (18) from [http-network-or-cache-fetch](https://fetch.spec.whatwg.org/#http-network-or-cache-fetch):

> 17. If httpRequest’s [cache mode](https://fetch.spec.whatwg.org/#concept-request-cache-mode) is "no-cache", httpRequest’s [prevent no-cache cache-control header modification flag](https://fetch.spec.whatwg.org/#no-cache-prevent-cache-control) is unset, and httpRequest’s [header list](https://fetch.spec.whatwg.org/#concept-request-header-list) [does not contain](https://fetch.spec.whatwg.org/#header-list-contains) `Cache-Control`, then [append](https://fetch.spec.whatwg.org/#concept-header-list-append) (`Cache-Control`, `max-age=0`) to httpRequest’s [header list](https://fetch.spec.whatwg.org/#concept-request-header-list).
> 18. If httpRequest’s [cache mode](https://fetch.spec.whatwg.org/#concept-request-cache-mode) is "no-store" or "reload", then:
>    1) If httpRequest’s [header list](https://fetch.spec.whatwg.org/#concept-request-header-list) [does not contain](https://fetch.spec.whatwg.org/#header-list-contains) `Pragma`, then [append](https://fetch.spec.whatwg.org/#concept-header-list-append) (`Pragma`, `no-cache`) to httpRequest’s [header list](https://fetch.spec.whatwg.org/#concept-request-header-list).
>    2) If httpRequest’s [header list](https://fetch.spec.whatwg.org/#concept-request-header-list) [does not contain](https://fetch.spec.whatwg.org/#header-list-contains) `Cache-Control`, then [append](https://fetch.spec.whatwg.org/#concept-header-list-append) (`Cache-Control`, `no-cache`) to httpRequest’s [header list](https://fetch.spec.whatwg.org/#concept-request-header-list).

**Comment by annevk** [comment]
@pmeenan would you be willing to drive this change?

**Comment by pmeenan** [comment]
Yep, happy to. Hopefully I learned some from the Priority Hints efforts and it will be a bit smoother.

Just to make sure we're all on the same page, the end state that we're planning to get to (once we get buy-in) is:

- [cache](https://fetch.spec.whatwg.org/#dom-request-cache) on Request will direct how the http cache treats requests.
- The cache attribute will not cause any headers to be modified (so impacting network caches requires manually setting the headers)
- [Specific Cache-control headers](https://fetch.spec.whatwg.org/#concept-request-cache-mode) will continue to cause cache to "no-store" if it is "default"

**Comment by annevk** [comment]
Yeah, that sounds right. In addition to what @mnot mentions above it would be good to double check all the text around "cache", including notes, to ensure it all makes sense still. Thank you!
